src/cargo/core/compiler/locking.rs

```diff
@@ -0,0 +1,224 @@
+//! This module handles the locking logic during compilation.
+//!
+//! The locking scheme is based on build unit level locking.
+//! Each build unit consists of a primary and secondary lock used to represent multiple lock states.
+//!
+//! | State                  | Primary     | Secondary   |
```

Comment:
Can we give semantic names to the locks, rather than using Primary and Secondary?